### DISCUSSION TITLE
How to apply/implement filtering on child resolvers? Not so trivial

### DISCUSSION THREAD

**Discussion by gabbigum** [graphql-java/graphql-java#2522] [Q&A]
For example, how can I fetch all the shows that have starRating of 3 following this schema?

```
type Query {
  shows: [Show]
}

type Show {
  showId: ID
  title: String
  reviews(minimumStarRating: Int): [Review]
}

type Review {
  starRating: Int
}
```

**The problem** is that we cannot remove a **_show_** from the parent (e.g. **_shows_**) when the show starRating is less than 3 stars leaving us with response like that:

```
# Response
"shows": [
      {
        "title": "Stranger Things",
        "reviews": [
          {
            "starRating": 5
          },
          {
            "starRating": 5
          },
          {
            "starRating": 4
          }
        ]
      },
      {
        "title": "Ozark", # This shouldn't be here 
        "reviews": null # because it doesn't meet my filtering criteria
      },
      {
        "title": "The Crown",
        "reviews": [
          {
            "starRating": 3
          },
          {
            "starRating": 4
          },
          {
            "starRating": 4
          }
        ]
      }
]
```
Thanks in advance!

**Comment by kvfedorov** [comment]
The way I read your schema is that it is not about filtering shows, but about filtering reviews. So it would allow to filter reviews of particular show, not set of shows themselves. I would add _another_ query to allow filter shows by review rating:

```
type Query {
    shows: [Show]
    showsByRatings(input: ShowsRatingFilter): [Show]
}

input ShowsRatingFilter {
    minimumRating: Int
    maximumRating: Int
}
...
```
Then, it should be trivial to implement resolver for `showsByRatings`.

Hope it helps

**Reply by gabbigum** [reply]
Thanks for the reply, but this solution doesn't fulfil my team goal. I used the shows and reviews as simplified version of what we are trying to achieve. It is too expensive for us to perform all the filtration in the query resolver, we want to be able to perform the filtration in the child resolvers, thats why we need to remove shows along the way when they don't match the filter criteria. Bottom line we should we able to return a response which doesn't have any null-s (e.g. match all the filtering criteria).

**Reply by kvfedorov** [reply]
Example schema you have allows to write query like this:
```
query {
    shows {
         reviews(minimumStarRating: 3) {
              starRating
         }
    }
}
```
I believe, in GraphQL it reads as: return ALL shows, for EACH show, return only reviews with minimal star rating 3. It is NOT a query to return only shows with rating 3.

So it is not needed nor possible to filter parents from inside child resolver.

**Reply by gabbigum** [reply]
The reason we want to perform query like this is because our entities are linked by foreign keys(strings with resource links) and not by object references due to internal framework reasons. That's why we need to somehow make it possible to filter from inside child resolvers. Thanks for the answer by the way!

**Reply by kvfedorov** [reply]
I'm not sure why underlying storage implementation would matter here. We use DynamoDb which does not have object references either and our keys are strings as well. It should be just a matter of schema and correspondent resolver you implement to bridge the gap between your storage and that schema.

**Reply by gabbigum** [reply]
I was talking about that our java objects are linked by strings and not by references 😃   sorry for not clarifying.

**Reply by kvfedorov** [reply]
As I said, it does not matter. GraphQL is agnostic to that. See my very first comment and move your query from child resolver to `showsByRating` resolver and it should work without any hacks.

**Comment by dugenkui03** [comment]
Here is my way: [graphql-calculator](https://github.com/dugenkui03/graphql-java-calculator/blob/main/src/test/java/calculator/engine/directive/FilterTest.java).

[Directives](https://spec.graphql.org/draft/#sec-Language.Directives) is a elegant way to satisfied your goal: filter element which not satisfied condition. There is the definition of @filter:
```graphql
directive @filter(predicate: String!) on FIELD
```

we can get the infomation about @filter and list field in Instrumentation and remove the element which not satisfied the condition. the query like that:
```
query($minimumStarRating:Int){
  shows
  @filter(predicate:"reviews == null")
  {
     showId
     title
     reviews(minimumStarRating: $minimumStarRating){
        starRating
     }
  }
}
```
More cases: [FilterTest.java](https://github.com/dugenkui03/graphql-java-calculator/blob/main/src/test/java/calculator/engine/directive/FilterTest.java).
